engine/derivative.cc

```diff
@@ -639,6 +640,53 @@ namespace MathDAG
        throw error("frac is not differentiable");
   }
   
+  template <>
+  NodePtr SystemOfEquations::derivative<>
+  (const OperationDAG<OperationType::percent>& expr)
+  {
+    if (expr.arguments[0].empty())
+      return zero;
+    else
+      {
+        Expr x(expressionCache, expr.arguments[0][0]);
+        return (100*x)->derivative(*this);
+      }
+  }
+
+  template <>
+  NodePtr SystemOfEquations::derivative
+  (const OperationDAG<OperationType::polygamma>& expr)
+  {
+    assert(expr.arguments.size()==2);
+    if (expr.arguments[0].empty())
+      return zero;
+    else if (expr.arguments[1].empty())
+      {
+        Expr x(expressionCache, expr.arguments[0][0]);
+        return chainRule(x,polygamma(one,x));
+      }
+    else
+      {
+        Expr x(expressionCache, expr.arguments[0][0]);
+        Expr y(expressionCache,  expr.arguments[1][0]);
+        return chainRule(x,polygamma(y+one,x));;
+      }
```

Comment:
This entails you're making polygamma a binary operation, with the second argument the order. But the second argument can change as a function of time, which would cause a lot of surprises. I'd prefer that the order were a separate attribute of the operation, that is fixed for all simulation time. You could potentially use the arg attribute of Operation for this purpose. It is a double, not an int, but you could check it for integralness at reset time.

Reply:
Can I still leave it as a binop as in the most recent commit? I've added it as a ConstantDAG in expr.h:

  inline Expr polygamma(const Expr& x, double y) {
    shared_ptr<OperationDAGBase> r(x.newNode(OperationType::polygamma));
    r->arguments[0].push_back(x);
    r->arguments[1].push_back(**NodePtr(new ConstantDAG(y))**);    
    return Expr(x.cache,r);
  }     

and also call it as polygamma(x,1+expr.arguments[1][0]) in derivative.cc now.

Reply:
OK - let's suck it and see.